File: cli/internal/broker/transport.go
// Package broker implements a lightweight WebSocket broker that bridges
// browser clients (Skuld protocol) with a Claude Code CLI process (SDK
// protocol). It can be embedded in the Forge server for mini mode or
// run as a standalone process for VM deployments.
//
// TODO(standalone): Add NewStandaloneTransport that spawns Claude CLI
// and manages its own SDK WebSocket server, enabling `niuu volundr broker`
// as a standalone command for non-k8s deployments.
package broker

// Transport abstracts CLI communication so the broker doesn't depend on
// Forge's SDKTransport directly. In embedded mode, Forge's SDKTransport
// implements this. In standalone mode, a transport that spawns and manages
// the CLI process directly would implement this.
type Transport interface {
	// SendUserMessage sends a user message to the CLI process.
	// content can be a string or an array of content blocks.
	SendUserMessage(content any, cliSessionID string) error

	// SendControlResponse sends a control_response envelope to the CLI.
	SendControlResponse(response map[string]any) error

	// CLISessionID returns the session ID reported by the CLI on init.
	CLISessionID() string
}

File: src/cli/broker/transport.py
"""Transport protocol — abstraction for CLI communication."""

from __future__ import annotations

from typing import Any, Protocol


class Transport(Protocol):
    """Abstraction for sending messages to the Claude Code CLI process.

    In embedded mode, Volundr's SDK transport implements this.
    In standalone mode, a transport that spawns and manages the CLI
    process directly would implement this.
    """

    def send_user_message(self, content: Any, cli_session_id: str) -> None:
        """Send a user message to the CLI process.

        ``content`` can be a string or an array of content blocks.
        """
        ...

    def send_control_response(self, response: dict[str, Any]) -> None:
        """Send a control_response envelope to the CLI."""
        ...

    def cli_session_id(self) -> str:
        """Return the session ID reported by the CLI on init."""
        ...
